controllers/cks_cluster_controller.go

```diff
@@ -0,0 +1,120 @@
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-cloudstack/api/v1beta3"
+	csCtrlrUtils "sigs.k8s.io/cluster-api-provider-cloudstack/controllers/utils"
+)
+
+const CksClusterFinalizer = "ckscluster.infrastructure.cluster.x-k8s.io"
+
+// RBAC permissions for CloudStackCluster.
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=cloudstackclusters,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=cloudstackclusters/status,verbs=create;get;update;patch
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=cloudstackclusters/finalizers,verbs=update
+
+// CksClusterReconciliationRunner is a ReconciliationRunner with extensions specific to CloudStackClusters.
+// The runner does the actual reconciliation.
+type CksClusterReconciliationRunner struct {
+	*csCtrlrUtils.ReconciliationRunner
+	FailureDomains        *infrav1.CloudStackFailureDomainList
+	ReconciliationSubject *infrav1.CloudStackCluster
+}
+
+// CksClusterReconciler is the k8s controller manager's interface to reconcile a CloudStackCluster.
+// This is primarily to adapt to k8s.
+type CksClusterReconciler struct {
+	csCtrlrUtils.ReconcilerBase
+}
+
+// Initialize a new CloudStackCluster reconciliation runner with concrete types and initialized member fields.
+func NewCksClusterReconciliationRunner() *CksClusterReconciliationRunner {
+	// Set concrete type and init pointers.
+	runner := &CksClusterReconciliationRunner{ReconciliationSubject: &infrav1.CloudStackCluster{}}
+	runner.FailureDomains = &infrav1.CloudStackFailureDomainList{}
+	// Setup the base runner. Initializes pointers and links reconciliation methods.
+	runner.ReconciliationRunner = csCtrlrUtils.NewRunner(runner, runner.ReconciliationSubject, "CKSClusterController")
+	runner.CSCluster = runner.ReconciliationSubject
+	return runner
+}
+
+// Reconcile is the method k8s will call upon a reconciliation request.
+func (reconciler *CksClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (retRes ctrl.Result, retErr error) {
+	r := NewCksClusterReconciliationRunner()
+	r.UsingBaseReconciler(reconciler.ReconcilerBase).ForRequest(req).WithRequestCtx(ctx)
+	r.WithAdditionalCommonStages(r.GetFailureDomains(r.FailureDomains))
+	return r.RunBaseReconciliationStages()
+}
+
+// Reconcile actually reconciles the CloudStackCluster.
+func (r *CksClusterReconciliationRunner) Reconcile() (res ctrl.Result, reterr error) {
+	if !r.CSCluster.Spec.SyncWithACS {
+		return ctrl.Result{}, nil
+	}
+	// Prevent premature deletion.
+	controllerutil.AddFinalizer(r.ReconciliationSubject, CksClusterFinalizer)
+
+	if len(r.FailureDomains.Items) == 0 {
+		return r.RequeueWithMessage("No failure domains found")
```

Comment:
should this be an error? or maybe a no-op?
Why do we reconcile? it seems like the user would need to make changes for this to start working, right? Maybe I'm missing something

Reply:
Makes sense. Let me do no-op here to match the behavior with cloudstack controller.